src/main/java/me/oyurimatheus/nossomercadolivre/products/NewProductRequest.java

```diff
@@ -0,0 +1,99 @@
+package me.oyurimatheus.nossomercadolivre.products;
+
+import me.oyurimatheus.nossomercadolivre.categories.Category;
+import me.oyurimatheus.nossomercadolivre.users.User;
+import org.hibernate.validator.constraints.Length;
+
+import javax.persistence.JoinColumn;
+import javax.persistence.ManyToOne;
+import javax.validation.constraints.DecimalMin;
+import javax.validation.constraints.Min;
+import javax.validation.constraints.NotBlank;
+import javax.validation.constraints.Size;
+import java.math.BigDecimal;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.function.Function;
+
+import static java.lang.String.format;
+import static java.util.stream.Collectors.toList;
+
+class NewProductRequest {
+
+    @NotBlank
+    private String name;
+
+    @DecimalMin(value = "0.01")
+    private BigDecimal price;
+
+    @Min(0)
+    private Integer stockQuantity;
+
+    /**
+     * Photos list in base 64
+     */
+    @Size(min = 1)
+    private List<String> photos;
+
+    @Size(min = 3)
+    private List<NewCharacteristicRequest> characteristics;
+
+    @NotBlank
+    @Length(max = 1000)
+    private String description;
+
+    @ManyToOne
+    @JoinColumn(name = "category_id")
+    private Long categoryId;
+
+    /**
+     * @deprecated frameworks eyes only
+     */
+    @Deprecated
+    private NewProductRequest() { }
+
+    public String getName() {
+        return name;
+    }
+
+    public BigDecimal getPrice() {
+        return price;
+    }
+
+    public Integer getStockQuantity() {
+        return stockQuantity;
+    }
+
+    public List<String> getPhotos() {
+        return photos;
+    }
+
+    public List<NewCharacteristicRequest> getCharacteristics() {
+        return characteristics;
+    }
+
+    public String getDescription() {
+        return description;
+    }
+
+    public Long getCategoryId() {
+        return categoryId;
+    }
+
+    public Product toProduct(PhotoUploader photoUploader, Function<Long, Optional<Category>> findCategoryById, User user) {
+
+        UUID productId = UUID.randomUUID();
```

Comment:
Eu não tinha visto aqui... sempre que a mesma variável é usada mais de uma vez dentro de um fluxo eu entendo como um smell de consistência. Esse uuid que você usou nas fotos é o que você quer no produto. Mas aqui nada te impede de usar um uuid para a foto e outro para a criação do produto.

Se o upload da foto tem tanta dependencia com o produto, talvez essa operação pudesse acontecer depois da criação... Talvez as fotos devem ser entidades mesmo.

Reply:
Percebe que você também o ```user``` duas vezes.

Reply:
Não sei dizer se as fotos seriam uma entity... Para mim elas têm uma dependência muito forte do produto. Não sei se faz sentido criar o produto e depois atribuir as fotos, porque um produto precisa de, pelo menos, uma foto para existir... O que acha de talvez criar uma `factory` ou um `builder` onde encapsula essa criação de produto e o upload das fotos?

Reply:
Eu entendo aqui, to bem aberto para qualquer solução que seja proporcional a dificuldade da feature. Qualquer coisa mais simples que não use a variável duas vezes, pode valer :). Essa é a restrição.